Index: xorg-server-1.7.3.902/hw/dmx/dmxinit.c
===================================================================
--- xorg-server-1.7.3.902.orig/hw/dmx/dmxinit.c	2009-12-25 15:28:41.000000000 -0800
+++ xorg-server-1.7.3.902/hw/dmx/dmxinit.c	2010-02-03 17:29:02.000000000 -0800
@@ -834,7 +834,7 @@
 /** This function is called in Xserver/os/utils.c from \a AbortServer().
  * We must ensure that backend and console state is restored in the
  * event the server shutdown wasn't clean. */
-void AbortDDX(void)
+void SigAbortDDX(int signo)
 {
     int i;
 
@@ -846,6 +846,11 @@
     }
 }
 
+void AbortDDX(void)
+{
+    SigAbortDDX(0);
+}
+
 /** This function is called in Xserver/dix/main.c from \a main() when
  * dispatchException & DE_TERMINATE (which is the only way to exit the
  * main loop without an interruption. */
Index: xorg-server-1.7.3.902/hw/kdrive/src/kdrive.c
===================================================================
--- xorg-server-1.7.3.902.orig/hw/kdrive/src/kdrive.c	2009-10-27 23:56:21.000000000 -0700
+++ xorg-server-1.7.3.902/hw/kdrive/src/kdrive.c	2010-02-03 17:29:02.000000000 -0800
@@ -322,7 +322,7 @@
 }
 
 void
-AbortDDX(void)
+SigAbortDDX(int signo)
 {
     KdDisableScreens ();
     if (kdOsFuncs)
@@ -339,6 +339,12 @@
 }
 
 void
+AbortDDX(void)
+{
+    SigAbortDDX(0);
+}
+
+void
 ddxGiveUp (void)
 {
     AbortDDX ();
Index: xorg-server-1.7.3.902/hw/vfb/InitOutput.c
===================================================================
--- xorg-server-1.7.3.902.orig/hw/vfb/InitOutput.c	2009-10-27 23:56:21.000000000 -0700
+++ xorg-server-1.7.3.902/hw/vfb/InitOutput.c	2010-02-03 17:29:02.000000000 -0800
@@ -214,11 +214,17 @@
 }
 
 void
-AbortDDX(void)
+SigAbortDDX(int signo)
 {
     ddxGiveUp();
 }
 
+void
+AbortDDX(void)
+{
+    SigAbortDDX(0);
+}
+
 #ifdef __APPLE__
 void
 DarwinHandleGUI(int argc, char *argv[])
Index: xorg-server-1.7.3.902/hw/xfree86/common/xf86Events.c
===================================================================
--- xorg-server-1.7.3.902.orig/hw/xfree86/common/xf86Events.c	2009-11-29 21:42:31.000000000 -0800
+++ xorg-server-1.7.3.902/hw/xfree86/common/xf86Events.c	2010-02-03 17:29:02.000000000 -0800
@@ -355,6 +355,8 @@
 int
 xf86SigWrapper(int signo)
 {
+  static Bool beenhere = FALSE;
+
   if ((signo == SIGILL) && xf86SigIllHandler) {
     (*xf86SigIllHandler)();
     return 0; /* continue */
Index: xorg-server-1.7.3.902/hw/xfree86/common/xf86Init.c
===================================================================
--- xorg-server-1.7.3.902.orig/hw/xfree86/common/xf86Init.c	2009-12-25 15:28:41.000000000 -0800
+++ xorg-server-1.7.3.902/hw/xfree86/common/xf86Init.c	2010-02-03 17:29:02.000000000 -0800
@@ -1174,14 +1174,17 @@
 }
 
 /*
- * ddxGiveUp --
+ * ddxSigGiveUp --
  *      Device dependent cleanup. Called by by dix before normal server death.
  *      For SYSV386 we must switch the terminal back to normal mode. No error-
  *      checking here, since there should be restored as much as possible.
+ *
+ *      If a non-zero signo is passed, re-raise that signal rather than
+ *      calling abort().
  */
 
 void
-ddxGiveUp(void)
+ddxSigGiveUp(int signo)
 {
     int i;
 
@@ -1207,24 +1210,45 @@
 
     xf86CloseConsole();
 
+    ErrorF (" ddxSigGiveUp: Closing log\n");
     xf86CloseLog();
 
     /* If an unexpected signal was caught, dump a core for debugging */
-    if (xf86Info.caughtSignal)
-	abort();
+    if (xf86Info.caughtSignal) {
+        if (signo != 0) {
+            raise(signo);
+        } else {
+            abort();
+        }
+    }
 }
 
+/*
+ * ddxGiveUp --
+ *      Device dependent cleanup. Called by by dix before normal server death.
+ *      For SYSV386 we must switch the terminal back to normal mode. No error-
+ *      checking here, since there should be restored as much as possible.
+ */
+
+void
+ddxGiveUp()
+{
+    ddxSigGiveUp(0);
+}
 
 
 /*
- * AbortDDX --
+ * SigAbortDDX --
  *      DDX - specific abort routine.  Called by AbortServer(). The attempt is
  *      made to restore all original setting of the displays. Also all devices
  *      are closed.
+ *
+ *      If a non-zero signo is passed, re-raise that signal rather than calling
+ *      abort()
  */
 
 void
-AbortDDX(void)
+SigAbortDDX(int signo)
 {
   int i;
 
@@ -1255,7 +1279,20 @@
    * This is needed for an abnormal server exit, since the normal exit stuff
    * MUST also be performed (i.e. the vt must be left in a defined state)
    */
-  ddxGiveUp();
+  ddxSigGiveUp(signo);
+}
+
+/*
+ * AbortDDX --
+ *      DDX - specific abort routine.  The attempt is made to restore
+ *      all original setting of the displays. Also all devices are
+ *      closed.
+ */
+
+void
+AbortDDX()
+{
+    SigAbortDDX(0);
 }
 
 void
Index: xorg-server-1.7.3.902/hw/xnest/Init.c
===================================================================
--- xorg-server-1.7.3.902.orig/hw/xnest/Init.c	2009-10-27 23:56:22.000000000 -0700
+++ xorg-server-1.7.3.902/hw/xnest/Init.c	2010-02-03 17:29:02.000000000 -0800
@@ -112,12 +112,17 @@
 /*
  * DDX - specific abort routine.  Called by AbortServer().
  */
-void AbortDDX(void)
+void SigAbortDDX(int signo)
 {
   xnestDoFullGeneration = True;
   xnestCloseDisplay();
 }
 
+void AbortDDX(void)
+{
+    SigAbortDDX(0);
+}
+
 /* Called by GiveUp(). */
 void ddxGiveUp(void)
 {
Index: xorg-server-1.7.3.902/hw/xquartz/darwin.c
===================================================================
--- xorg-server-1.7.3.902.orig/hw/xquartz/darwin.c	2009-10-27 23:56:22.000000000 -0700
+++ xorg-server-1.7.3.902/hw/xquartz/darwin.c	2010-02-03 17:29:02.000000000 -0800
@@ -798,12 +798,12 @@
 
 
 /*
- * AbortDDX --
+ * [Sig]AbortDDX --
  *      DDX - specific abort routine.  Called by AbortServer(). The attempt is
  *      made to restore all original setting of the displays. Also all devices
  *      are closed.
  */
-void AbortDDX( void )
+void SigAbortDDX( void )
 {
     ErrorF( "   AbortDDX\n" );
     /*
@@ -813,6 +813,11 @@
     ddxGiveUp();
 }
 
+void AbortDDX( void )
+{
+    SigAbortDDX(0);
+}
+
 #include "mivalidate.h" // for union _Validate used by windowstr.h
 #include "windowstr.h"  // for struct _Window
 #include "scrnintstr.h" // for struct _Screen
Index: xorg-server-1.7.3.902/hw/xwin/InitOutput.c
===================================================================
--- xorg-server-1.7.3.902.orig/hw/xwin/InitOutput.c	2009-12-25 15:28:42.000000000 -0800
+++ xorg-server-1.7.3.902/hw/xwin/InitOutput.c	2010-02-03 17:29:02.000000000 -0800
@@ -283,7 +283,7 @@
 
 /* See Porting Layer Definition - p. 57 */
 void
-AbortDDX (void)
+SigAbortDDX (int signo)
 {
 #if CYGDEBUG
   winDebug ("AbortDDX\n");
@@ -291,6 +291,12 @@
   ddxGiveUp ();
 }
 
+void
+AbortDDX (void)
+{
+    SigAbortDDX(0);
+}
+
 #ifdef __CYGWIN__
 /* hasmntopt is currently not implemented for cygwin */
 static const char *winCheckMntOpt(const struct mntent *mnt, const char *opt)
Index: xorg-server-1.7.3.902/include/os.h
===================================================================
--- xorg-server-1.7.3.902.orig/include/os.h	2009-12-25 15:28:42.000000000 -0800
+++ xorg-server-1.7.3.902/include/os.h	2010-02-03 17:29:02.000000000 -0800
@@ -434,7 +434,9 @@
 /* stuff for FlushCallback */
 extern _X_EXPORT CallbackListPtr FlushCallback;
 
+extern _X_EXPORT void SigAbortDDX(int signo);
 extern _X_EXPORT void AbortDDX(void);
+extern _X_EXPORT void ddxSigGiveUp(int signo);
 extern _X_EXPORT void ddxGiveUp(void);
 extern _X_EXPORT int TimeSinceLastInputEvent(void);
 
Index: xorg-server-1.7.3.902/os/log.c
===================================================================
--- xorg-server-1.7.3.902.orig/os/log.c	2009-12-25 15:28:42.000000000 -0800
+++ xorg-server-1.7.3.902/os/log.c	2010-02-03 17:29:30.000000000 -0800
@@ -389,11 +389,12 @@
 }
 
 #ifdef __GNUC__
+void SigAbortServer(int signo) __attribute__((noreturn));
 void AbortServer(void) __attribute__((noreturn));
 #endif
 
 void
-AbortServer(void)
+SigAbortServer(int signo)
 {
 #ifdef XF86BIGFONT
     XF86BigfontCleanup();
@@ -401,11 +402,21 @@
     CloseWellKnownConnections();
     OsCleanup(TRUE);
     CloseDownDevices();
-    AbortDDX();
+    SigAbortDDX(signo);
     fflush(stderr);
-    if (CoreDump)
-	abort();
-    exit (1);
+    if (CoreDump) {
+        if (signo != 0)
+            raise(signo);
+        else
+            abort();
+    } else
+	exit (1);
+}
+
+void
+AbortServer(void)
+{
+    SigAbortServer(0);
 }
 
 #define AUDIT_PREFIX "AUDIT: %s: %ld: "
@@ -508,6 +519,27 @@
 }
 
 void
+FatalSignal(int signo)
+{
+    static Bool beenhere = FALSE;
+
+    if (beenhere)
+	ErrorF("\nFatalSignal re-entered, aborting\n");
+    else
+	ErrorF("\nCaught signal %d (%s). Server aborting\n",
+               signo, strsignal(signo));
+
+    if (!beenhere)
+	OsVendorFatalError();
+    if (!beenhere) {
+	beenhere = TRUE;
+	SigAbortServer(signo);
+    } else
+	abort();
+    /*NOTREACHED*/
+}
+
+void
 FatalError(const char *f, ...)
 {
     va_list args;
Index: xorg-server-1.7.3.902/os/osinit.c
===================================================================
--- xorg-server-1.7.3.902.orig/os/osinit.c	2009-10-27 23:56:24.000000000 -0700
+++ xorg-server-1.7.3.902/os/osinit.c	2010-02-03 17:29:02.000000000 -0800
@@ -148,13 +148,13 @@
           case SIGBUS:
           case SIGILL:
           case SIGFPE:
+	      signal(signo,SIG_DFL);
 	      ErrorF("%s at address %p\n", strsignal(signo), sip->si_addr);
       }
   }
 #endif
 
-  FatalError("Caught signal %d (%s). Server aborting\n",
-	     signo, strsignal(signo));
+  FatalSignal(signo);
 }
 
 void
