Index: xserver-xorg-input-synaptics/man/synaptics.man
===================================================================
--- xserver-xorg-input-synaptics.orig/man/synaptics.man
+++ xserver-xorg-input-synaptics/man/synaptics.man
@@ -871,15 +871,6 @@
 32 bit unsigned, 2 values (read-only), vertical, horizontal in units/millimeter.
 
 .SH "NOTES"
-There is an example hal policy file in
-.I ${sourcecode}/fdi/11-x11-synaptics.fdi
-which will enable the driver based on the information if the hardware
-is available. Feel free to copy it to
-.I /etc/hal/fdi/policy
-and customize it to your needs. You can pass custom options to the driver
-using x11_options properties. Note that this requires xorg-server-1.5 or
-higher.
-.LP
 If either of
 .BI "Protocol \*q" auto-dev \*q
 (default) or
